File: yua-sdk-python/src/yua/_streaming.py
from __future__ import annotations

import json
import math
from typing import (
    Any,
    Callable,
    Dict,
    Generic,
    Iterator,
    List,
    Optional,
    TypeVar,
)

import httpx

from yua.types.chat import (
    ChatCompletion,
    ChatCompletionChoice,
    ChatCompletionChunk,
    ChatCompletionMessage,
    CompletionUsage,
    YuaExtension,
)

T = TypeVar("T")


class Stream(Generic[T]):
    """
    Stream<T> — Iterator + .on() event handlers + convenience methods.

    Backend stream lifecycle:
      READY -> STREAMING (tokens) -> FINAL (logical end) -> DONE (transport close)
    """

    def __init__(
        self,
        response: httpx.Response,
        parse: Callable[[Dict[str, Any]], Optional[T]],
    ) -> None:
        self._response = response
        self._parse = parse
        self._listeners: Dict[str, List[Callable[..., Any]]] = {}

        # eventId reorder buffer
        self._last_event_id = -1
        self._pending: Dict[int, Dict[str, Any]] = {}

        # Aggregated YUA data
        self._activities: List[Any] = []
        self._suggestions: List[Any] = []
        self._memory_ops: List[Any] = []
        self._reasoning_blocks: List[Any] = []

    def on(self, event: str, handler: Callable[..., Any]) -> "Stream[T]":
        self._listeners.setdefault(event, []).append(handler)
        return self

    def _emit(self, event: str, data: Any) -> None:
        for handler in self._listeners.get(event, []):
            try:
                handler(data)
            except Exception:
                pass

    def close(self) -> None:
        self._response.close()

    def __iter__(self) -> Iterator[T]:
        try:
            for frame in _parse_sse(self._response):
                event_id = frame.get("eventId")
                if isinstance(event_id, int):
                    if event_id <= self._last_event_id:
                        continue
                    if event_id != self._last_event_id + 1 and self._last_event_id >= 0:
                        self._pending[event_id] = frame
                        continue
                    self._last_event_id = event_id

                result = self._process_frame(frame)
                if result is not None:
                    yield result

                # Flush buffered frames
                while (self._last_event_id + 1) in self._pending:
                    nxt = self._pending.pop(self._last_event_id + 1)
                    self._last_event_id += 1
                    r = self._process_frame(nxt)
                    if r is not None:
                        yield r
        finally:
            self._emit("done", None)
            self._response.close()

    def _process_frame(self, frame: Dict[str, Any]) -> Optional[T]:
        parsed = self._parse(frame)
        if parsed is None:
            return None

        # Aggregate YUA extension data + emit events
        chunk: Any = parsed
        if hasattr(chunk, "yua_event") and chunk.yua_event is not None:
            evt_type = chunk.yua_event.type
            evt_data = chunk.yua_event.data
            self._emit(evt_type, evt_data)

            if evt_type == "activity" and evt_data.get("activity", {}).get("item"):
                activity = evt_data["activity"]
                op = activity.get("op")
                item = activity["item"]
                if op == "ADD":
                    self._activities.append(item)
                elif op == "PATCH":
                    for i, a in enumerate(self._activities):
                        if isinstance(a, dict) and a.get("id") == item.get("id"):
                            self._activities[i] = {**a, **item}
                            break

            if evt_type == "suggestion" and evt_data.get("suggestions"):
                self._suggestions = evt_data["suggestions"]

            if evt_type == "memory" and evt_data.get("memory"):
                self._memory_ops.append(evt_data["memory"])

            if evt_type == "reasoning_block" and evt_data.get("block"):
                self._reasoning_blocks.append(evt_data["block"])

        return parsed

    def text_content(self) -> str:
        text = ""
        for chunk in self:
            c: Any = chunk
            if hasattr(c, "choices") and c.choices:
                delta = c.choices[0].delta
                if hasattr(delta, "content") and delta.content is not None:
                    text += delta.content
        return text

    def final_message(self) -> ChatCompletion:
        last_id = ""
        full_text = ""
        model = ""
        finish_reason = "stop"

        for chunk in self:
            c: Any = chunk
            if hasattr(c, "id") and c.id:
                last_id = c.id
            if hasattr(c, "model") and c.model:
                model = c.model
            if hasattr(c, "choices") and c.choices:
                delta = c.choices[0].delta
                if hasattr(delta, "content") and delta.content is not None:
                    full_text += delta.content
                fr = c.choices[0].finish_reason
                if fr:
                    finish_reason = fr

        return ChatCompletion(
            id=last_id,
            object="chat.completion",
            created=int(math.floor(___now())),
            model=model,
            choices=[
                ChatCompletionChoice(
                    index=0,
                    message=ChatCompletionMessage(role="assistant", content=full_text),
                    finish_reason=finish_reason,
                )
            ],
            usage=CompletionUsage(),
            yua=YuaExtension(
                activities=self._activities,
                suggestions=self._suggestions,
                memory_ops=self._memory_ops,
                reasoning_blocks=self._reasoning_blocks,
            ),
        )


def ___now() -> float:
    import time
    return time.time()


def _parse_sse(response: httpx.Response) -> Iterator[Dict[str, Any]]:
    """Parse SSE stream from httpx response."""
    buffer = ""
    for text_chunk in response.iter_text():
        buffer += text_chunk

        while "\n\n" in buffer:
            boundary = buffer.index("\n\n")
            raw_event = buffer[:boundary]
            buffer = buffer[boundary + 2 :]

            event_type: Optional[str] = None
            data_lines: List[str] = []

            for line in raw_event.split("\n"):
                if line.startswith("event:"):
                    event_type = line[6:].strip()
                elif line.startswith("data:"):
                    data_lines.append(line[5:].strip())

            if not data_lines:
                continue

            data = "\n".join(data_lines)

            # Skip heartbeat / keep-alive
            if data in ("[heartbeat]", "") or data.startswith(":"):
                continue
            # SSE end signal
            if data == "[DONE]":
                return

            try:
                frame = json.loads(data)
            except json.JSONDecodeError:
                continue

            # done=true means transport close
            if frame.get("done") is True:
                return

            # Inject event type into frame for mapper
            if event_type:
                frame["_sse_event"] = event_type

            yield frame

File: yua-sdk/src/core/streaming.ts
import type { ChatCompletion, ChatCompletionChunk, ActivityItem, SuggestionItem, MemoryPayload, ReasoningBlock } from "../types/chat";

type EventHandler<T> = (data: T) => void;

/**
 * Stream<T> — AsyncIterable + .on() event handlers + convenience methods.
 *
 * Backend stream lifecycle:
 *   READY → STREAMING (tokens) → FINAL (logical end) → DONE (transport close)
 *
 * FINAL = no more tokens, UI can stop cursor
 * DONE  = SSE connection closed, cleanup
 *
 * Events are ordered by monotonic `eventId` assigned by StreamEngine.
 */
export class Stream<T = ChatCompletionChunk>
  implements AsyncIterable<T>
{
  private readonly body: ReadableStream<Uint8Array>;
  private readonly controller: AbortController;
  private readonly parse: (raw: SSEFrame) => T | null;

  private listeners: Map<string, EventHandler<any>[]> = new Map();

  // eventId reorder buffer (backend assigns monotonic eventId per session)
  private lastEventId = -1;
  private pendingFrames: Map<number, SSEFrame> = new Map();

  // Aggregated YUA data for finalMessage()
  private _activities: ActivityItem[] = [];
  private _suggestions: SuggestionItem[] = [];
  private _memoryOps: MemoryPayload[] = [];
  private _reasoningBlocks: ReasoningBlock[] = [];

  constructor(
    body: ReadableStream<Uint8Array>,
    controller: AbortController,
    parse: (raw: SSEFrame) => T | null,
  ) {
    this.body = body;
    this.controller = controller;
    this.parse = parse;
  }

  on(event: "stage", handler: EventHandler<{ stage: string }>): this;
  on(event: "activity", handler: EventHandler<{ activity: any }>): this;
  on(event: "reasoning_block", handler: EventHandler<{ block: ReasoningBlock }>): this;
  on(event: "reasoning_done", handler: EventHandler<void>): this;
  on(event: "suggestion", handler: EventHandler<{ suggestions: SuggestionItem[] }>): this;
  on(event: "memory", handler: EventHandler<{ memory: MemoryPayload }>): this;
  on(event: "answer_unlocked", handler: EventHandler<void>): this;
  on(event: "done", handler: EventHandler<void>): this;
  on(event: string, handler: EventHandler<any>): this {
    const list = this.listeners.get(event) ?? [];
    list.push(handler);
    this.listeners.set(event, list);
    return this;
  }

  private emit(event: string, data: unknown): void {
    const list = this.listeners.get(event);
    if (!list) return;
    for (const handler of list) {
      try {
        handler(data);
      } catch {
        // swallow listener errors
      }
    }
  }

  abort(): void {
    this.controller.abort();
  }

  async *[Symbol.asyncIterator](): AsyncIterator<T> {
    for await (const frame of parseSSE(this.body)) {
      // eventId-based ordering: if out of order, buffer and replay
      const eventId = frame.json?.eventId;
      if (typeof eventId === "number") {
        if (eventId <= this.lastEventId) continue; // dedup

        if (eventId !== this.lastEventId + 1 && this.lastEventId >= 0) {
          // Out of order — buffer it
          this.pendingFrames.set(eventId, frame);
          continue;
        }

        this.lastEventId = eventId;
      }

      const result = this.processFrame(frame);
      if (result) yield result;

      // Flush any buffered frames that are now in order
      while (this.pendingFrames.has(this.lastEventId + 1)) {
        const next = this.pendingFrames.get(this.lastEventId + 1)!;
        this.pendingFrames.delete(this.lastEventId + 1);
        this.lastEventId++;
        const r = this.processFrame(next);
        if (r) yield r;
      }
    }

    this.emit("done", undefined);
  }

  private processFrame(frame: SSEFrame): T | null {
    const parsed = this.parse(frame);
    if (!parsed) return null;

    // Aggregate YUA extension data + emit events
    const chunk = parsed as any;
    if (chunk.yua_event) {
      const { type, data } = chunk.yua_event;
      this.emit(type, data);

      // Aggregate for finalMessage()
      if (type === "activity" && data?.activity?.item) {
        const { op, item } = data.activity;
        if (op === "ADD") this._activities.push(item);
        else if (op === "PATCH") {
          const idx = this._activities.findIndex((a) => a.id === item.id);
          if (idx >= 0) Object.assign(this._activities[idx], item);
        }
      }
      if (type === "suggestion" && data?.suggestions) {
        this._suggestions = data.suggestions;
      }
      if (type === "memory" && data?.memory) {
        this._memoryOps.push(data.memory);
      }
      if (type === "reasoning_block" && data?.block) {
        this._reasoningBlocks.push(data.block);
      }
    }

    return parsed;
  }

  async textContent(): Promise<string> {
    let text = "";
    for await (const chunk of this) {
      const c = chunk as any;
      const delta = c.choices?.[0]?.delta?.content;
      if (typeof delta === "string") {
        text += delta;
      }
    }
    return text;
  }

  async finalMessage(): Promise<ChatCompletion> {
    let lastId = "";
    let fullText = "";
    let model = "";
    let finishReason = "stop";
    let thinkingProfile: string | undefined;

    for await (const chunk of this) {
      const c = chunk as any;
      if (c.id) lastId = c.id;
      if (c.model) model = c.model;
      const delta = c.choices?.[0]?.delta?.content;
      if (typeof delta === "string") fullText += delta;
      const fr = c.choices?.[0]?.finish_reason;
      if (fr) finishReason = fr;
      // Extract thinking profile from meta
      if (c.yua_event?.type === "stage" && !thinkingProfile) {
        thinkingProfile = c.meta?.thinkingProfile;
      }
    }

    return {
      id: lastId,
      object: "chat.completion",
      created: Math.floor(Date.now() / 1000),
      model,
      choices: [
        {
          index: 0,
          message: { role: "assistant", content: fullText },
          finish_reason: finishReason as any,
        },
      ],
      usage: { prompt_tokens: 0, completion_tokens: 0, total_tokens: 0 },
      yua: {
        thinking_profile: (thinkingProfile as any) ?? undefined,
        activities: this._activities,
        suggestions: this._suggestions,
        memory_ops: this._memoryOps,
        reasoning_blocks: this._reasoningBlocks,
      },
    };
  }
}

// --- SSE Parser (from legacy yua-one-node, battle-tested) ---

export interface SSEFrame {
  event?: string;
  data: string;
  json?: any;
}

async function* parseSSE(
  body: ReadableStream<Uint8Array>,
): AsyncGenerator<SSEFrame> {
  const reader = body.getReader();
  const decoder = new TextDecoder("utf-8");
  let buffer = "";

  try {
    while (true) {
      const { value, done } = await reader.read();
      if (done) break;
      buffer += decoder.decode(value, { stream: true });

      let boundary: number;
      while ((boundary = buffer.indexOf("\n\n")) >= 0) {
        const rawEvent = buffer.slice(0, boundary);
        buffer = buffer.slice(boundary + 2);

        const lines = rawEvent.split("\n");
        let eventType: string | undefined;
        const dataLines: string[] = [];

        for (const line of lines) {
          if (line.startsWith("event:")) {
            eventType = line.slice(6).trim();
          } else if (line.startsWith("data:")) {
            dataLines.push(line.slice(5).trim());
          }
        }

        if (!dataLines.length) continue;
        const data = dataLines.join("\n");

        // Skip heartbeat / keep-alive (backend sends 15s ping)
        if (data === "[heartbeat]" || data === "" || data.startsWith(":")) continue;
        // SSE end signal
        if (data === "[DONE]") return;

        let json: any;
        try {
          json = JSON.parse(data);
        } catch {
          // not JSON — skip
          continue;
        }

        // done=true means transport close — stop parsing
        if (json.done === true) return;

        yield { event: eventType, data, json };
      }
    }
  } finally {
    reader.releaseLock();
  }
}
